File: sam.cc
#include "sam.h"
#include "types.h"
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <unordered_map>
#include <utility>

int size;
int next;
block *sam;
bool *used;
std::unordered_map<int, std::pair<block *, size_t>> cache;

int reads = 0;
int writes = 0;
int allocs = 0;

void init(int n) {
  size = n;
  reads = 0;
  writes = 0;
  allocs = 0;

  next = 0;
  sam = new block[n];
  used = new bool[n];
  memset(used, 0, n * sizeof(bool));
}

void clear() {
  size = 0;
  delete[] sam;
  delete[] used;
}

void debug_header() {
  std::cout << "READS" << "\t" << "WRITES" << "\t" << "ALLOCS" << '\n';
}

void debug() { std::cout << reads << "\t" << writes << "\t" << allocs << '\n'; }

void debug_reset() {
  reads = 0;
  writes = 0;
  allocs = 0;
}

void write(int i, block x) {
  ++writes;
  sam[i] = x;
}

block read(int i) {
  ++reads;
  if (used[i]) {
    int x = -1;
    std::cerr << "ADDRESS " << i << " ALREADY READ!\n";
    std::cerr << "SAM SIZE " << size << '\n';
    std::cerr << "ALLOCATIONS " << allocs << '\n';
    std::exit(1);
  } else {
    used[i] = true;
    return sam[i];
  }
}

int alloc() {
  ++allocs;
  return next++;
}

void link(int l, int r, int p) {
  write(l, {(int)block_type::INNER, p, r});
  write(r, {(int)block_type::INNER, p, l});
}

std::pair<int, int> copy(int p) {
  int l = alloc();
  int r = alloc();
  link(l, r, p);
  return {l, r};
}

int copy_cache(int* p) {
  int l = alloc();
  int r = alloc();
  link(l, r, *p);
  *p = r;
  return l;
}

void destroy(int x) {
  if (cache.contains(x)) {
    return;
  }
  const auto bl = read(x);
  if ((block_type)bl[0] == block_type::DATA) {
    // DO NOTHING
  } else {
    const auto p = bl[1];
    const auto s = bl[2];

    if (cache.contains(p)) {
      int dummy = alloc();
      write(dummy, {(int)block_type::INNER, p, s});
      write(s, {(int)block_type::INNER, p, dummy});
      return;
    }

    const auto bl2 = read(p);
    if ((block_type)bl2[0] == block_type::DATA) {
      write(s, bl2);
    } else {
      const auto p_ = bl2[1];
      const auto s_ = bl2[2];
      link(s, s_, p_);
    }
  }
}

void destroy_cache(int x) {

}

/**
 * Chase pointer x, splaying the tree as it goes.
 * As input take (1) a pointer x and (2) two elements that should point to the
 * new version of x.
 *
 * Return (1) the pointer where pointed-to data should be written back and (2)
 * the two fields of the pointee.
 */

int counter = 0;
std::pair<int, block> splay(int a, int b, int x) {
  while (true) {
    ++counter;

    const auto bl = read(x);
    const auto tx = (block_type)bl[0];
    const auto y = bl[1];
    const auto c = bl[2];

    if (tx == block_type::DATA) {
      int x_ = alloc();
      link(a, b, x_);
      return {x_, bl};
    }

    const auto bl2 = read(y);
    const auto ty = (block_type)bl2[0];
    const auto z = bl2[1];
    const auto d = bl2[2];
    if (ty == block_type::DATA) {
      /**
       * zig splay:
       *
       *     y            x
       *    / \          / \
       *   x   c  ==>   a   y
       *  / \              / \
       * a   b            b   c
       */

      int x_ = alloc();
      int y_ = alloc();

      link(b, c, y_);
      link(a, y_, x_);

      return {x_, bl2};
    }

    /**
     * zig-zag splay:
     *       z               x
     *      / \             / \
     *     y   d           /   \
     *    / \      ==>    y     z
     *   c   x           / \   / \
     *      / \         a   b c   d
     *     a   b
     */

    int x_ = z;
    int y_ = alloc();
    int z_ = alloc();

    link(a, b, y_);
    link(c, d, z_);

    a = y_;
    b = z_;
    x = x_;
  }
}

// splays tree, but returns from cache at last level if hit
// else puts into cache
// in either case, returns the writeback address/cache tag
int counter_cache = 0;
CacheObj splay_cache(int a, int b, int x) {
  while (true) {
    ++counter;

    if (cache.contains(x)) {
      link(a, b, x);
      return CacheObj(x);
    }

    auto bl = read(x);
    const auto tx = (block_type)bl[0];
    const auto y = bl[1];
    const auto c = bl[2];

    if (tx == block_type::DATA) {
      int x_ = alloc();
      link(a, b, x_);

      // put data into the cache
      bl[15] = x_;
      cache[x_] = {new block(bl), 0};
      return CacheObj(x_);
    }

    if (cache.contains(y)) {
      int x_ = alloc();
      link(a, b, x_);
      link(x_, c, y);

      return CacheObj(y);
    }

    auto bl2 = read(y);
    const auto ty = (block_type)bl2[0];
    const auto z = bl2[1];
    const auto d = bl2[2];
    if (ty == block_type::DATA) {
      /**
       * zig splay:
       *
       *     y            x
       *    / \          / \
       *   x   c  ==>   a   y
       *  / \              / \
       * a   b            b   c
       */

      int x_ = alloc();
      int y_ = alloc();

      link(b, c, y_);
      link(a, y_, x_);

      bl2[15] = x_;
      cache[x_] = {new block(bl2), 0};
      return CacheObj(x_);
    }

    /**
     * zig-zag splay:
     *       z               x
     *      / \             / \
     *     y   d           /   \
     *    / \      ==>    y     z
     *   c   x           / \   / \
     *      / \         a   b c   d
     *     a   b
     */

    int x_ = z;
    int y_ = alloc();
    int z_ = alloc();

    link(a, b, y_);
    link(c, d, z_);

    a = y_;
    b = z_;
    x = x_;
  }
}

std::tuple<int, int, block> deref(int x) {
  const auto bl = read(x);
  const auto t = (block_type)bl[0];
  const auto p = bl[1];
  const auto s = bl[2];

  const auto x_ = alloc();
  if (t == block_type::DATA) {
    return {x_, x_, bl};
  } else {
    const auto [where, b] = splay(x_, s, p);
    return {x_, where, b};
  }
}

CacheObj deref_cache(int *x) {
  if (cache.contains(*x)) {
    return CacheObj(*x);
  }

  auto bl = read(*x);
  const auto t = (block_type)bl[0];
  const auto p = bl[1];
  const auto s = bl[2];

  const auto x_ = alloc();
  *x = x_;

  if (t == block_type::DATA) {
    cache[x_] = {new block(bl), 0};
    return CacheObj(x_);
  } else {
    return splay_cache(x_, s, p);
  }
}

void cache_evict(int x) {
  if (!cache.contains(x)) {
    return;
  }
  auto data = cache.at(x);
  write(x, *(data.first));
  delete data.first;
  cache.erase(x);
}

CacheObj deref_cache(int *x);

block make_data_block() {
  block b;
  b[0] = (int)block_type::DATA;
  for (int i = 1; i < b.size(); i++) {
    b[i] = -1;
  }
  return b;
}


CacheObj::CacheObj() {
  int addr = alloc();
  block b = make_data_block();

  cache[addr] = {new block(b), 1}; // one reference
  addr_ = addr;
}

// Default constructor
CacheObj::CacheObj(int addr) : addr_{addr} { cache.at(addr).second += 1; }

// Copy constructor
CacheObj::CacheObj(const CacheObj &other) : addr_(other.addr_) {
  cache.at(addr_).second += 1;
}

// Copy assignment
CacheObj &CacheObj::operator=(const CacheObj &other) {
  if (this != &other) {
    // Decrement old reference
    release();

    // Copy from other
    addr_ = other.addr_;

    // Increment new reference
    cache.at(addr_).second += 1;
  }
  return *this;
}

int CacheObj::at(int idx) {
  assert(idx < 4);
  block *b = cache.at(addr_).first;
  return (*b)[idx];
}

void CacheObj::set(int idx, int val) {
  block *b = cache.at(addr_).first;
  (*b)[idx] = val;
}

CachePtr CacheObj::ptr_at(int idx) {
  return CachePtr(addr_, idx);
}

// Destructor
CacheObj::~CacheObj() { release(); }

void CacheObj::release() {
  cache.at(addr_).second -= 1;
  if (cache.at(addr_).second == 0) {
    cache_evict(addr_);
  }
}

// Default constructor
CachePtr::CachePtr(int addr, int idx) : addr_{addr}, idx_{idx} {
  cache.at(addr).second += 1;
}

// Copy constructor
CachePtr::CachePtr(const CachePtr &other) : addr_(other.addr_), idx_(other.idx_) {
  cache.at(addr_).second += 1;
}

CachePtr &CachePtr::operator=(const CachePtr &other) {
  if (this != &other) {
    // Decrement old reference
    cache.at(addr_).second -= 1;
    if (cache.at(addr_).second == 0) {
      cache_evict(addr_);
    }

    // Copy from other
    addr_ = other.addr_;
    idx_ = other.idx_;

    // Increment new reference
    cache.at(addr_).second += 1;
  }
  return *this;
}

CacheObj CachePtr::deref() {
  block *b = cache.at(addr_).first;
  return deref_cache(&(*b)[idx_ + 4]);
}

void CachePtr::set(CachePtr other) {
  destroy();
  block *b1 = cache.at(addr_).first;
  block* b2 = cache.at(other.addr_).first;
  int* p1 = &(*b1)[idx_ + 4];
  int* p2 = &(*b2)[other.idx_ + 4];
  *p1 = copy_cache(p2);
}

CacheObj CachePtr::alloc_object() {
  destroy();
  block *b1 = cache.at(addr_).first;
  int* p1 = &(*b1)[idx_ + 4];
  int a = alloc();
  *p1 = a;
  block b = make_data_block();
  cache[a] = {new block(b), 0};
  return CacheObj(a);
}

void CachePtr::destroy() {
  block *b = cache.at(addr_).first;
  int* p = &(*b)[idx_ + 4];
  if (*p == -1) {
    // slot is empty
    return;
  }
  destroy_cache(*p); // handle case where slot is not empty
}

// Destructor
CachePtr::~CachePtr() {
  cache.at(addr_).second -= 1;
  if (cache.at(addr_).second == 0) {
    cache_evict(addr_);
  }
}

File: python-impl/sam.py
from typing import Tuple, Dict, Optional, List
from sam_types import Block, BlockType, make_block, BLOCK_SIZE
import sys

class SAM:
    def __init__(self):
        self.size = 0
        self.next = 0
        self.sam: Dict[int, Block] = {}
        self.used: set[int] = set()
        self.cache: Dict = {}

        # Statistics
        self.reads = 0
        self.writes = 0
        self.allocs = 0

    def init(self, n: int):
        """Initialize SAM with n slots"""
        self.size = n
        self.reads = 0
        self.writes = 0
        self.allocs = 0
        self.next = 0
        # Use a dictionary instead of a list to avoid preallocating huge arrays
        self.sam = {}
        self.used = set()  # Use a set to track used addresses
        self.cache = {}

    def clear(self):
        """Clear the SAM"""
        self.size = 0
        self.sam = {}
        self.used = set()
        self.cache = {}

    def debug_reset(self):
        """Reset SAM usage counters"""
        self.reads = 0
        self.writes = 0
        self.allocs = 0

    def debug_header(self):
        print("READS\tWRITES\tALLOCS")

    def debug(self):
        """Print SAM usage"""
        print(f"{self.reads}\t{self.writes}\t{self.allocs}")

    def alloc(self) -> int:
        """Set aside a pointer with no pointee"""
        self.allocs += 1
        result = self.next
        self.next += 1
        return result

    def write(self, p: int, b: Block):
        """Write block b to pointer p"""
        self.writes += 1
        self.sam[p] = b.copy()

    def read(self, i: int) -> Block:
        """Read from address i (can only be done once per address)"""
        self.reads += 1
        if i in self.used:
            print(f"ADDRESS {i} ALREADY READ!", file=sys.stderr)
            print(f"SAM SIZE {self.size}", file=sys.stderr)
            print(f"ALLOCATIONS {self.allocs}", file=sys.stderr)
            sys.exit(1)
        else:
            self.used.add(i)
            return self.sam[i].copy()

    def link(self, l: int, r: int, p: int):
        """Create linked INNER blocks"""
        self.write(l, [BlockType.INNER.value, p, r] + [-1] * (BLOCK_SIZE - 3))
        self.write(r, [BlockType.INNER.value, p, l] + [-1] * (BLOCK_SIZE - 3))

    def copy(self, p: int) -> Tuple[int, int]:
        """
        Copy pointer p, invalidating p but yielding two fresh pointers p' and p''
        """
        l = self.alloc()
        r = self.alloc()
        self.link(l, r, p)
        return (l, r)

    def destroy(self, x: int):
        """Destroy pointer x"""
        if x in self.cache:
            return

        bl = self.read(x)
        if BlockType(bl[0]) == BlockType.DATA:
            # DO NOTHING
            pass
        else:
            p = bl[1]
            s = bl[2]

            if p in self.cache:
                dummy = self.alloc()
                self.write(dummy, [BlockType.INNER.value, p, s] + [-1] * (BLOCK_SIZE - 3))
                self.write(s, [BlockType.INNER.value, p, dummy] + [-1] * (BLOCK_SIZE - 3))
                return

            bl2 = self.read(p)
            if BlockType(bl2[0]) == BlockType.DATA:
                self.write(s, bl2)
            else:
                p_ = bl2[1]
                s_ = bl2[2]
                self.link(s, s_, p_)

    def splay(self, a: int, b: int, x: int) -> Tuple[int, Block]:
        """
        Chase pointer x, splaying the tree as it goes.
        Return the writeback address and the block data.
        """
        while True:
            bl = self.read(x)
            tx = BlockType(bl[0])
            y = bl[1]
            c = bl[2]

            if tx == BlockType.DATA:
                x_ = self.alloc()
                self.link(a, b, x_)
                return (x_, bl)

            bl2 = self.read(y)
            ty = BlockType(bl2[0])
            z = bl2[1]
            d = bl2[2]

            if ty == BlockType.DATA:
                # zig splay
                x_ = self.alloc()
                y_ = self.alloc()

                self.link(b, c, y_)
                self.link(a, y_, x_)

                return (x_, bl2)

            # zig-zag splay
            x_ = z
            y_ = self.alloc()
            z_ = self.alloc()

            self.link(a, b, y_)
            self.link(c, d, z_)

            a = y_
            b = z_
            x = x_

    def deref(self, x: int) -> Tuple[int, int, Block]:
        """
        Dereference pointer x yielding block b.
        Returns [p', q, b] where:
        - p' can be used to replace p
        - q is where to write back an updated version of b
        - b is the block data
        """
        bl = self.read(x)
        t = BlockType(bl[0])
        p = bl[1]
        s = bl[2]

        x_ = self.alloc()
        if t == BlockType.DATA:
            return (x_, x_, bl)
        else:
            where, b = self.splay(x_, s, p)
            return (x_, where, b)

    def save(self, p: int, data: Block) -> int:
        """Helper function to dereference, then write back data"""
        p_, where, b = self.deref(p)
        self.write(where, data)
        return p_

    def cache_evict(self, x: int):
        """Evict address x from cache and write it back to SAM"""
        if x not in self.cache:
            return
        block_ptr, refcount = self.cache[x]
        self.write(x, block_ptr)
        del self.cache[x]

    def splay_cache(self, a: int, b: int, x: int):
        """
        Splay operation that returns a CacheObj.
        Like splay(), but returns cached objects and checks cache at each step.
        """
        while True:
            if x in self.cache:
                self.link(a, b, x)
                return CacheObj(x)

            bl = self.read(x)
            tx = BlockType(bl[0])
            y = bl[1]
            c = bl[2]

            if tx == BlockType.DATA:
                x_ = self.alloc()
                self.link(a, b, x_)

                # put data into the cache
                bl[15] = x_
                self.cache[x_] = [bl, 0]  # [block, refcount]
                return CacheObj(x_)

            if y in self.cache:
                x_ = self.alloc()
                self.link(a, b, x_)
                self.link(x_, c, y)
                return CacheObj(y)

            bl2 = self.read(y)
            ty = BlockType(bl2[0])
            z = bl2[1]
            d = bl2[2]

            if ty == BlockType.DATA:
                # zig splay
                x_ = self.alloc()
                y_ = self.alloc()

                self.link(b, c, y_)
                self.link(a, y_, x_)

                bl2[15] = x_
                self.cache[x_] = [bl2, 0]
                return CacheObj(x_)

            # zig-zag splay
            x_ = z
            y_ = self.alloc()
            z_ = self.alloc()

            self.link(a, b, y_)
            self.link(c, d, z_)

            a = y_
            b = z_
            x = x_

    def deref_cache(self, block_ref: Block, idx: int):
        """
        Dereference a pointer stored in block_ref[idx], updating it in place.
        Returns a CacheObj.
        """
        x = block_ref[idx]

        if x in self.cache:
            return CacheObj(x)

        bl = self.read(x)
        t = BlockType(bl[0])
        p = bl[1]
        s = bl[2]

        x_ = self.alloc()
        block_ref[idx] = x_  # Update the pointer in place

        if t == BlockType.DATA:
            self.cache[x_] = [bl, 0]
            return CacheObj(x_)
        else:
            return self.splay_cache(x_, s, p)

# Global SAM instance
_sam = SAM()

# Expose module-level functions
def init(n: int):
    _sam.init(n)

def clear():
    _sam.clear()

def debug_reset():
    _sam.debug_reset()

def debug_header():
    _sam.debug_header()

def debug():
    _sam.debug()

def alloc() -> int:
    return _sam.alloc()

def write(p: int, b: Block):
    _sam.write(p, b)

def read(i: int) -> Block:
    return _sam.read(i)

def copy(p: int) -> Tuple[int, int]:
    return _sam.copy(p)

def destroy(x: int):
    _sam.destroy(x)

def deref(x: int) -> Tuple[int, int, Block]:
    return _sam.deref(x)

def save(p: int, data: Block) -> int:
    return _sam.save(p, data)

# Cache-related classes and functions

class CacheObj:
    """
    Wrapper for cached blocks with automatic reference counting.
    When all references are released, the block is evicted from cache.
    """
    def __init__(self, addr: Optional[int] = None):
        if addr is None:
            # Default constructor - allocate new object
            self.addr_ = _sam.alloc()
            b = [BlockType.DATA.value] + [-1] * (BLOCK_SIZE - 1)
            _sam.cache[self.addr_] = [b, 1]  # refcount starts at 1
        else:
            # Constructor from address
            self.addr_ = addr
            _sam.cache[self.addr_][1] += 1  # Increment refcount

    def __del__(self):
        self.release()

    def destroy(self):
        pass

    def release(self):
        """Explicitly decrement reference count and evict if zero (call instead of relying on __del__)"""
        if self.addr_ is not None and self.addr_ in _sam.cache:
            _sam.cache[self.addr_][1] -= 1
            if _sam.cache[self.addr_][1] == 0:
                _sam.cache_evict(self.addr_)
            self.addr_ = None  # Prevent double-destroy

    def at(self, idx: int) -> int:
        """Read value at index (for first 4 fields of the block)"""
        assert idx < 4, f"Index {idx} out of range, must be < 4"
        block_ref = _sam.cache[self.addr_][0]
        return block_ref[idx]

    def set(self, idx: int, val: int):
        """Write value at index (for first 4 fields of the block)"""
        block_ref = _sam.cache[self.addr_][0]
        block_ref[idx] = val

    def deref_at(self, idx: int):
        """Dereference pointer at field idx+4, returns new CacheObj"""
        block_ref = _sam.cache[self.addr_][0]
        return _sam.deref_cache(block_ref, idx + 4)

    def __copy__(self):
        """Copy constructor"""
        raise ValueError

    def __deepcopy__(self, memo):
        """Deep copy - same as shallow copy for CacheObj"""
        raise ValueError

    def ptr_at(self, idx: int):
        """Get a CachePtr pointing to field idx+4 of this object"""
        return CachePtr(self.addr_, idx)

class CachePtr:
    def __init__(self, addr: int, idx: int):
        self.addr_ = addr
        self.idx_ = idx
        _sam.cache[addr][1] += 1  # Increment refcount of parent CacheObj

    def destroy(self):
        pass

    def __del__(self):
        self.release()

    def release(self):
        """Explicitly release reference to parent CacheObj (call instead of relying on __del__)"""
        if self.addr_ is not None and self.addr_ in _sam.cache:
            _sam.cache[self.addr_][1] -= 1
            if _sam.cache[self.addr_][1] == 0:
                _sam.cache_evict(self.addr_)
            self.addr_ = None  # Prevent double-destroy
    
    def deref(self) -> CacheObj:
        """Dereference this pointer, returning the pointed-to CacheObj"""
        block_ref = _sam.cache[self.addr_][0]
        return _sam.deref_cache(block_ref, self.idx_ + 4)

    def set(self, other: 'CachePtr'):
        """
        Set this pointer to point to the same thing as other.
        Calls copy_cache to duplicate the pointer.
        """
        self.destroy_slot()
        b1 = _sam.cache[self.addr_][0]
        b2 = _sam.cache[other.addr_][0]

        # p2 points to b2[other.idx_ + 4]
        # We need to call copy_cache which takes a pointer and modifies it
        # copy_cache(p2) will modify b2[other.idx_ + 4] and return a new pointer

        # Simulate copy_cache(&(*b2)[other.idx_ + 4])
        p_val = b2[other.idx_ + 4]
        l = _sam.alloc()
        r = _sam.alloc()
        _sam.link(l, r, p_val)
        b2[other.idx_ + 4] = r  # Modify b2's pointer

        # Set p1 to the returned value
        b1[self.idx_ + 4] = l

    def alloc_object(self) -> CacheObj:
        """
        Allocate a new CacheObj and make this pointer point to it.
        """
        self.destroy_slot()
        block_ref = _sam.cache[self.addr_][0]

        a = _sam.alloc()
        block_ref[self.idx_ + 4] = a

        b = [BlockType.DATA.value] + [-1] * (BLOCK_SIZE - 1)
        _sam.cache[a] = [b, 0]
        return CacheObj(a)

    def destroy_slot(self):
        """Destroy the pointer stored at this location (if any)"""
        block_ref = _sam.cache[self.addr_][0]
        p = block_ref[self.idx_ + 4]
        if p == -1:
            # slot is empty
            return
        # destroy_cache is a no-op in the C++ code, so we do nothing
